pkg/storaged/btrfs/subvolume.jsx

```diff
@@ -44,19 +47,140 @@ function subvolume_mount(volume, subvol, forced_options) {
+                          passno: { t: 'i', v: 0 },
+                          "track-parents": { t: 'b', v: true }
+                      }
+                  ];
+
+    return block.AddConfigurationItem(config, {})
+            .then(reload_systemd)
+            .then(() => {
+                if (mount_now) {
+                    return client.mount_at(block, mount_point);
+                } else
+                    return Promise.resolve();
+            });
+}
+
+function subvolume_create(volume, subvol, parent_dir) {
+    const block = client.blocks[volume.path];
+
+    const action_variants = [
+        { tag: null, Title: _("Create and mount") },
+        { tag: "nomount", Title: _("Create only") }
+    ];
+
+    dialog_open({
+        Title: _("Create subvolume"),
+        Fields: [
+            TextInput("name", _("Name"),
+                      {
+                          validate: name => validate_subvolume_name(name)
+                      }),
+            TextInput("mount_point", _("Mount Point"),
+                      {
+                          validate: val => {
+                              // Mount points are optional for subvolumes
+                              if (val !== "")
+                                  return is_valid_mount_point(client, block, client.add_mount_point_prefix(val), false, true, subvol);
+                          }
+                      }),
+            mount_options(false, false),
+            at_boot_input("local"),
+        ],
+        Action: {
+            Variants: action_variants,
+            action: function (vals) {
+                if (vals.mount_point === "" && vals.variant !== "nomount")
+                    return Promise.reject(_("Mount point required for create and mount"));
+                // HACK: cannot use block_btrfs.CreateSubvolume as it always creates a subvolume relative to MountPoints[0] which
+                // makes it impossible to handle a situation where we have multiple subvolumes mounted.
+                // https://github.com/storaged-project/udisks/issues/1242
+                return cockpit.spawn(["btrfs", "subvolume", "create", `${parent_dir}/${vals.name}`], { superuser: "require", err: "message" }).then(() => {
+                    btrfs_poll();
+                    if (vals.mount_point !== "") {
+                        return set_mount_options(subvol, block, vals);
+                    }
+                });
+            }
+        }
+    });
+}
+
 export function make_btrfs_subvolume_page(parent, volume, subvol) {
     const actions = [];
 
     const use = btrfs_usage(client, volume);
     const block = client.blocks[volume.path];
     const fstab_config = get_fstab_config_with_client(client, block, false, subvol);
-    const [, mount_point] = fstab_config;
+    const [, mount_point, opts] = fstab_config;
+    const opt_ro = extract_option(parse_options(opts), "ro");
+    if (subvol.pathname === "quiche/pizza/bonk") {
+        console.log(subvol);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19849-20240117-115707-49793047-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)